modules/roles.js

```diff
@@ -0,0 +1,198 @@
+}
+/** @param {Discord.GuildMember} member */
+function getInventory(member) {
+  return u.perms.isMgmt(member) ? equipRoles : equipRoles.filter(r => member.roles.cache.hasAny(r.baseId, ...r.inherited));
+}
+
+/** @param {Augur.GuildInteraction<"CommandSlash">} int */
+async function slashRoleInventory(int) {
+  try {
+    const member = int.member;
+    const inventory = getInventory(member);
+    const inv = int.guild.roles.cache
+      .filter(r => inventory.has(r.id))
+      .sort((a, b) => b.comparePositionTo(a))
+      .map(r => r);
+    const embed = u.embed({ author: member })
+      .setTitle("Equippable Color Inventory")
+      .setDescription(`Equip a color role with </role equip:${u.sf.commands.slashRole}>\n\n${inv.join("\n")}`);
+    if (inv.length == 0) int.reply({ content: "You don't have any colors in your inventory!", ephemeral: true });
+    else int.reply({ embeds: [embed], ephemeral: int.channel?.id != u.sf.channels.botspam });
+  } catch (e) { u.errorHandler(e, int); }
+}
+
+/** @param {Augur.GuildInteraction<"CommandSlash">} int */
+async function slashRoleEquip(int) {
+  try {
+    await int.deferReply({ ephemeral: int.channel?.id != u.sf.channels.botspam });
+    const allColors = equipRoles.map(r => r.colorId);
+    const available = getInventory(int.member);
+    const input = int.options.getString("color");
+    if (!input) {
+      await int.member.roles.remove(allColors);
+      return int.editReply("Color removed!");
+    }
+    const real = int.guild.roles.cache.find(r => r.name.toLowerCase() == input.toLowerCase());
+    const role = int.guild.roles.cache.find(r => r.id == equipRoles.find(a => a.baseId == real?.id || a.colorId == real?.id)?.colorId);
+    if (!role) {
+      return int.editReply(`Sorry, that's not a color role on this server. Check </role inventory:${u.sf.commands.slashRole}> to see what you can equip.`);
+    } else if (!available.has(role.id)) {
+      return int.editReply(`Sorry, you don't have that color in your inventory. Check </role inventory:${u.sf.commands.slashRole}> to see what you can equip.`);
+    } else {
+      await int.member.roles.remove(allColors);
+      await int.member.roles.add(role.id);
+      int.editReply("Color applied!");
+    }
+  } catch (e) { u.errorHandler(e, int); }
+}
+
+async function setRoles() {
+  try {
+    const filterBlank = (input) => input != "" && input != null && input != undefined;
+    const ldsg = Module.client.guilds.cache.get(u.sf.ldsg);
+    if (!ldsg) throw new Error("LDSG is unavailable");
+    const doc = new GoogleSpreadsheet(config.google.sheets.config);
+    await doc.useServiceAccountAuth(config.google.creds);
+    await doc.loadInfo();
+    /** @type {{RoleId: string}[]} */
+    // @ts-ignore
+    const rows = await doc.sheetsByTitle["Opt-In Roles"].getRows();
+    optRoles.clear();
+    const ids = rows.map(r => r["RoleId"]).filter(filterBlank);
+    optRoles = ldsg?.roles.cache.filter(r => ids.includes(r.id)) ?? new u.Collection();
```

Comment:
ldsg wouldn't be undefined at this point

```suggestion
    optRoles = ldsg.roles.cache.filter(r => ids.includes(r.id));
```